notebooks/official/workbench/spark/spark_sample_notebook.ipynb

```diff
@@ -0,0 +1,1200 @@
+    "\n",
+    "You can create combinations of languages in each repo using the original data fetched from BigQuery."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "25479dcebd93"
+   },
+   "outputs": [],
+   "source": [
+    "# A Python package to get combinations.\n",
+    "from itertools import combinations\n",
+    "\n",
+    "# A Python package to use type hint\n",
+    "from typing import List\n",
+    "\n",
+    "# PySpark functions\n",
+    "from pyspark.sql.functions import explode"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "ad6a3e548fc3"
+   },
+   "outputs": [],
+   "source": [
+    "def normalize_name(name: str) -> str:\n",
+    "    \"\"\"\n",
+    "    Change the language name to avoid invalid characters in the BigQuery data.\n",
+    "    Args:\n",
+    "        name: string\n",
+    "    Returns:\n",
+    "        Normalized name: string\n",
+    "    \"\"\"\n",
+    "    normalized_arr = []\n",
+    "\n",
+    "    # The following set of characters cannot be used in BigQuery's field.\n",
+    "    invalid_chars = {\",\", \";\", \"{\", \"}\", \"(\", \")\", \"\\n\", \"\\t\", \"=\"}\n",
+    "\n",
+    "    # The name must start with a letter or underscore.\n",
+    "    if name[0].isnumeric():\n",
+    "        normalized_arr.append(\"_\")\n",
+    "\n",
+    "    for ch in name:\n",
+    "        # Skip if a character is in the set of invalid characters.\n",
+    "        if ch in invalid_chars:\n",
+    "            continue\n",
+    "\n",
+    "        # Convert a space or dot to underscore\n",
+    "        elif ch == \" \" or ch == \".\":\n",
+    "            normalized_arr.append(\"_\")\n",
+    "\n",
+    "        # Change to lowercase to merge name duplications (for example, \"Java\" and \"java\")\n",
+    "        else:\n",
+    "            normalized_arr.append(ch.lower())\n",
+    "\n",
+    "    # Convert the array to string\n",
+    "    return \"\".join(normalized_arr)\n",
+    "\n",
+    "\n",
+    "@udf(returnType=ArrayType(StringType()))\n",
+    "def reduce_language(language) -> List[str]:\n",
+    "    \"\"\"\n",
+    "    The preprocess function takes the language and reduces it to remove \"bytes\".\n",
+    "    Args:\n",
+    "        language: list of struct that contains name and bytes.\n",
+    "                  (e.g., language = [[name: \"C\", bytes: 300],\n",
+    "                                     [name: \"Java\", bytes: 200]]\n",
+    "    Returns:\n",
+    "        list of strings that contains name.\n",
+    "                  (e.g., reduced_languages = [\"C\", \"Java\"])\n",
+    "    \"\"\"\n",
+    "    if len(language) < 2:\n",
+    "        return None\n",
+    "    reduced_languages = []\n",
+    "    for elem in language:\n",
+    "        # To write back to BigQuery, the name must be normalized (see the \"normalize_name()\" function below).\n",
```

Comment:
delete "(see the \"normalize_name()\" function below)" since it follows immediately.